core/src/commonMain/kotlin/com/xebia/functional/xef/llm/ChatWithFunctions.kt

```diff
@@ -31,8 +36,29 @@ suspend fun <A> ChatApi.prompt(
   scope: Conversation,
   serializer: KSerializer<A>,
 ): A =
-  prompt(prompt, scope, chatFunction(serializer.descriptor)) { json ->
-    Json.decodeFromString(serializer, json)
+  prompt(prompt, scope, chatFunctions(listOf(serializer.descriptor))) { call ->
+    Json.decodeFromString(serializer, call.arguments)
+  }
+
+@OptIn(ExperimentalSerializationApi::class)
+@AiDsl
+suspend fun <A> ChatApi.prompt(
+  prompt: Prompt<CreateChatCompletionRequestModel>,
+  scope: Conversation,
+  serializer: KSerializer<A>,
+  descriptors: List<SerialDescriptor>,
+): A =
+  prompt(prompt, scope, chatFunctions(descriptors)) { call ->
+    // adds a `type` field with the call.functionName serial name equivalent to the call arguments
+    val jsonWithDiscriminator = Json.decodeFromString(JsonElement.serializer(), call.arguments)
+    val descriptor =
+      descriptors.firstOrNull { it.serialName.endsWith(call.functionName) }
+        ?: error("No descriptor found for ${call.functionName}")
+    val newJson =
+      JsonObject(
+        jsonWithDiscriminator.jsonObject + ("type" to JsonPrimitive(descriptor.serialName))
```

Comment:
This addition supports polymorphic types like sealed classes whose serializer is chosen by the `type` property, the discriminator.